dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/reactive/AbstractTripleReactorPublisher.java

```diff
@@ -0,0 +1,190 @@
+
+    public AbstractTripleReactorPublisher() {
+    }
+
+    public AbstractTripleReactorPublisher(Consumer<CallStreamObserver<?>> onSubscribe, Runnable shutdownHook) {
+        this.onSubscribe = onSubscribe;
+        this.shutdownHook = shutdownHook;
+    }
+
+    protected void onSubscribe(final CallStreamObserver<?> subscription) {
+        if (subscription != null && SUBSCRIPTION.compareAndSet(this, null, subscription)) {
+            this.subscription = subscription;
+            if (subscription instanceof ClientStreamObserver<?>) {
+                ((ClientStreamObserver<?>) subscription).disableAutoRequest();
+            } else if (subscription instanceof ServerStreamObserver<?>) {
+                ((ServerStreamObserver<?>) subscription).disableAutoInboundFlowControl();
+            }
+            if (onSubscribe != null) {
+                onSubscribe.accept(subscription);
+            }
+            return;
+        }
+
+        throw new IllegalStateException(getClass().getSimpleName() + " supports only a single subscription");
+    }
+
+    @Override
+    public void onNext(T data) {
+        if (isDone || isCancelled) {
+            return;
+        }
+        downstream.onNext(data);
+    }
+
+    @Override
+    public void onError(Throwable throwable) {
+        if (isDone || isCancelled) {
+            return;
+        }
+        downstream.onError(throwable);
+        isDone = true;
+        doPostShutdown();
+    }
+
+    @Override
+    public void onCompleted() {
+        if (isDone || isCancelled) {
+            return;
+        }
+        isDone = true;
+        downstream.onComplete();
+        doPostShutdown();
+    }
+
+    private void doPostShutdown() {
+        Runnable r = shutdownHook;
+        // CAS to confirm shutdownHook will be run only once.
+        if (r != null && SHUTDOWN_HOOK.compareAndSet(this, r, null)) {
+            r.run();
+        }
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super T> subscriber) {
+        if (subscriber == null) {
+            throw new NullPointerException();
+        }
+
+        if (SUBSCRIBED.compareAndSet(false, true)) {
+            subscriber.onSubscribe(this);
+            this.downstream = subscriber;
+            if (isCancelled) {
+                this.downstream = null;
+            }
+        } else {
+            subscriber.onSubscribe(EMPTY_SUBSCRIPTION);
+            subscriber.onError(new IllegalStateException(getClass().getSimpleName() + " can't be subscribed repeatedly"));
+        }
+    }
+
+    @Override
+    public void request(long l) {
+        if (SUBSCRIBED.get() && CAN_REQUEST.get()) {
+            subscription.request(l >= Integer.MAX_VALUE ? Integer.MAX_VALUE : (int) l);
+        } else {
+            REQUESTED.getAndAdd(this, l);
+        }
+    }
+
+    @Override
+    public void startRequest() {
+        if (CAN_REQUEST.compareAndSet(false, true)) {
+            long count = requested;
```

Comment:
Is it possible to call request directly？

Reply:
No, in the scenario where the client uses serverStream, the request can only be made after ObserverToClientCallListenerAdapter#onStart is called, otherwise the NPE of TripleClientStream#request will be triggered, because the deframer of TripleClientStream has not been initialized.